In [1]:
import pandas as pd
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt

In [49]:
df_08 = pd.read_csv('all_alpha_08.csv')
df_18 = pd.read_csv('all_alpha_18.csv')

In [52]:
df_18.query('Fuel == "Electricity"')

Model  Displ  Cyl   Trans Drive         Fuel Cert Region  \
996  KIA Soul Electric    NaN  NaN  Auto-1   2WD  Electricity          FA   
997  KIA Soul Electric    NaN  NaN  Auto-1   2WD  Electricity          CA   

     Stnd      Stnd Description  Underhood ID      Veh Class  \
996  T3B0  Federal Tier 3 Bin 0  JKMXV00.0411  station wagon   
997   ZEV        California ZEV  JKMXV00.0411  station wagon   

     Air Pollution Score City MPG Hwy MPG Cmb MPG  Greenhouse Gas Score  \
996                   10      124      93     108                    10   
997                   10      124      93     108                    10   

    SmartWay Comb CO2  
996    Elite        0  
997    Elite        0

## Number of columns per dataset

In [3]:
df_08.info()
df_18.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2404 entries, 0 to 2403
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Model                 2404 non-null   object 
 1   Displ                 2404 non-null   float64
 2   Cyl                   2205 non-null   object 
 3   Trans                 2205 non-null   object 
 4   Drive                 2311 non-null   object 
 5   Fuel                  2404 non-null   object 
 6   Sales Area            2404 non-null   object 
 7   Stnd                  2404 non-null   object 
 8   Underhood ID          2404 non-null   object 
 9   Veh Class             2404 non-null   object 
 10  Air Pollution Score   2404 non-null   object 
 11  FE Calc Appr          2205 non-null   object 
 12  City MPG              2205 non-null   object 
 13  Hwy MPG               2205 non-null   object 
 14  Cmb MPG               2205 non-null   object 
 15  Unadj Cmb MPG        

## Number of records per dataset

In [ ]:
see above

## Number of duplicate rows

In [4]:
print(f'{df_08.duplicated().sum()} duplicated rows in the 08 dataset.')
print(f'{df_18.duplicated().sum()} duplicated rows in the 18 dataset.')

25 duplicated rows in the 08 dataset.
0 duplicated rows in the 18 dataset.


## Features with missing values

In [5]:
df_08.isnull().sum()

Model                     0
Displ                     0
Cyl                     199
Trans                   199
Drive                    93
Fuel                      0
Sales Area                0
Stnd                      0
Underhood ID              0
Veh Class                 0
Air Pollution Score       0
FE Calc Appr            199
City MPG                199
Hwy MPG                 199
Cmb MPG                 199
Unadj Cmb MPG           199
Greenhouse Gas Score    199
SmartWay                  0
dtype: int64

## Non-null unique values

In [6]:
df_08.nunique()

Model                   436
Displ                    47
Cyl                       8
Trans                    14
Drive                     2
Fuel                      5
Sales Area                3
Stnd                     12
Underhood ID            343
Veh Class                 9
Air Pollution Score      13
FE Calc Appr              2
City MPG                 39
Hwy MPG                  43
Cmb MPG                  38
Unadj Cmb MPG           721
Greenhouse Gas Score     20
SmartWay                  2
dtype: int64

## Count of unique fuel types

In [7]:
df_08.Fuel.nunique()

5

## Drop unneeded columns

## Rename 'Sales Area' column in 2008 dataframe

In [169]:
df_08

Model  Displ      Cyl    Trans Drive      Fuel Cert_Region Stnd  \
0       ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline          CA   U2   
1       ACURA MDX    3.7  (6 cyl)  Auto-S5   4WD  Gasoline          FA   B5   
2       ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline          CA   U2   
3       ACURA RDX    2.3  (4 cyl)  Auto-S5   4WD  Gasoline          FA   B5   
4        ACURA RL    3.5  (6 cyl)  Auto-S5   4WD  Gasoline          CA   U2   
...           ...    ...      ...      ...   ...       ...         ...  ...   
2399  VOLVO XC 90    3.2  (6 cyl)  Auto-S6   4WD  Gasoline          CA   U2   
2400  VOLVO XC 90    4.4  (8 cyl)  Auto-S6   4WD  Gasoline          CA   U2   
2401  VOLVO XC 90    3.2  (6 cyl)  Auto-S6   2WD  Gasoline          FA   B5   
2402  VOLVO XC 90    3.2  (6 cyl)  Auto-S6   4WD  Gasoline          FA   B5   
2403  VOLVO XC 90    4.4  (8 cyl)  Auto-S6   4WD  Gasoline          FA   B5   

      Underhood_ID    Veh_Class Air_Pollution_Score FE_Calc_Appr City_MPG  \
0     8HNXT03.7PKR          SUV                   7          Drv       15   
1     8HNXT03.7PKR          SUV                   6          Drv       15   
2     8HNXT02.3DKR          SUV                   7          Drv       17   
3     8HNXT02.3DKR          SUV                   6          Drv       17   
4     8HNXV03.5HKR  midsize car                   7          Drv       16   
...            ...          ...                 ...          ...      ...   
2399  8VVXB03.2U2N          SUV                   7          Drv       14   
2400  8VVXB04.4U2N          SUV                   7          Drv       13   
2401  8VVXB03.2U2N          SUV                   6          Drv       14   
2402  8VVXB03.2U2N          SUV                   6          Drv       14   
2403  8VVXB04.4U2N          SUV                   6          Drv       13   

     Hwy_MPG Cmb_MPG  Unadj_Cmb_MPG Greenhouse_Gas_Score SmartWay  
0         20      17        22.0527                    4       no  
1         20      17        22.0527                    4       no  
2         22      19        24.1745                    5       no  
3         22      19        24.1745                    5       no  
4         24      19        24.5629                    5       no  
...      ...     ...            ...                  ...      ...  
2399      20      16        20.7134                    4       no  
2400      19      15        19.4789                    3       no  
2401      20      16        20.9644                    4       no  
2402      20      16        20.7134                    4       no  
2403      19      15        19.4789                    3       no  

[2404 rows x 18 columns]

In [8]:
df_18

Model  Displ  Cyl       Trans Drive                  Fuel  \
0       ACURA RDX    3.5  6.0  SemiAuto-6   2WD              Gasoline   
1       ACURA RDX    3.5  6.0  SemiAuto-6   2WD              Gasoline   
2       ACURA RDX    3.5  6.0  SemiAuto-6   4WD              Gasoline   
3       ACURA RDX    3.5  6.0  SemiAuto-6   4WD              Gasoline   
4       ACURA TLX    2.4  4.0       AMS-8   2WD              Gasoline   
...           ...    ...  ...         ...   ...                   ...   
1606  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
1607  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
1608  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
1609  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
1610  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   

     Cert Region       Stnd            Stnd Description  Underhood ID  \
0             FA     T3B125      Federal Tier 3 Bin 125  JHNXT03.5GV3   
1             CA         U2      California LEV-II ULEV  JHNXT03.5GV3   
2             FA     T3B125      Federal Tier 3 Bin 125  JHNXT03.5GV3   
3             CA         U2      California LEV-II ULEV  JHNXT03.5GV3   
4             CA  L3ULEV125  California LEV-III ULEV125  JHNXV02.4WH3   
...          ...        ...                         ...           ...   
1606          FA      T3B70       Federal Tier 3 Bin 70  JVVXJ02.0A70   
1607          CA   L3ULEV70   California LEV-III ULEV70  JVVXJ02.0B70   
1608          FA      T3B70       Federal Tier 3 Bin 70  JVVXJ02.0B70   
1609          CA  L3SULEV30  California LEV-III SULEV30  JVVXJ02.0P30   
1610          FA      T3B30       Federal Tier 3 Bin 30  JVVXJ02.0P30   

         Veh Class  Air Pollution Score City MPG Hwy MPG Cmb MPG  \
0        small SUV                    3       20      28      23   
1        small SUV                    3       20      28      23   
2        small SUV                    3       19      27      22   
3        small SUV                    3       19      27      22   
4        small car                    3       23      33      27   
...            ...                  ...      ...     ...     ...   
1606  standard SUV                    5       22      28      24   
1607  standard SUV                    5       20      27      23   
1608  standard SUV                    5       20      27      23   
1609  standard SUV                    7    26/63   30/61   27/62   
1610  standard SUV                    7    26/63   30/61   27/62   

      Greenhouse Gas Score SmartWay Comb CO2  
0                        5       No      386  
1                        5       No      386  
2                        4       No      402  
3                        4       No      402  
4                        6       No      330  
...                    ...      ...      ...  
1606                     5       No      365  
1607                     5       No      394  
1608                     5       No      394  
1609                    10    Elite      187  
1610                    10    Elite      187  

[1611 rows x 18 columns]

In [9]:
df_08 = df_08.rename(columns=lambda x:x.replace(' ', '_'))
df_18 = df_18.rename(columns=lambda x:x.replace(' ', '_'))
df_08 = df_08.rename(columns={'Sales_Area' : 'Cert_Region', })

In [10]:

df_08.Cyl = df_08.Cyl.str.extract('(\d+)')
df_08 = df_08.drop(columns=['FE_Calc_Appr', 'Unadj_Cmb_MPG', 'SmartWay'])
df_18 = df_18.drop(columns=['Stnd_Description', 'SmartWay', 'Comb_CO2'])
df_08 = df_08.query('Cert_Region == "CA"')
df_18 = df_18.query('Cert_Region == "CA"')


## Make more consistent column labels

In [ ]:
see above

# Compare columns names and make sure they're the same

In [ ]:
see above

# CHECKPOINT: SAVE EVERYTHING THUS FAR INTO A DATAFRAME

## Only return records with a 'CA' cert_region

In [11]:
df_08

Model  Displ Cyl    Trans Drive      Fuel Cert_Region Stnd  \
0       ACURA MDX    3.7   6  Auto-S5   4WD  Gasoline          CA   U2   
2       ACURA RDX    2.3   4  Auto-S5   4WD  Gasoline          CA   U2   
4        ACURA RL    3.5   6  Auto-S5   4WD  Gasoline          CA   U2   
6        ACURA TL    3.2   6  Auto-S5   2WD  Gasoline          CA   U2   
7        ACURA TL    3.5   6  Auto-S5   2WD  Gasoline          CA   U2   
...           ...    ...  ..      ...   ...       ...         ...  ...   
2394    VOLVO V70    3.2   6  Auto-S6   2WD  Gasoline          CA   U2   
2396  VOLVO XC 70    3.2   6  Auto-S6   4WD  Gasoline          CA   U2   
2398  VOLVO XC 90    3.2   6  Auto-S6   2WD  Gasoline          CA   U2   
2399  VOLVO XC 90    3.2   6  Auto-S6   4WD  Gasoline          CA   U2   
2400  VOLVO XC 90    4.4   8  Auto-S6   4WD  Gasoline          CA   U2   

      Underhood_ID      Veh_Class Air_Pollution_Score City_MPG Hwy_MPG  \
0     8HNXT03.7PKR            SUV                   7       15      20   
2     8HNXT02.3DKR            SUV                   7       17      22   
4     8HNXV03.5HKR    midsize car                   7       16      24   
6     8HNXV03.5HKR    midsize car                   7       18      26   
7     8HNXV03.5RKR    midsize car                   7       17      26   
...            ...            ...                 ...      ...     ...   
2394  8VVXB03.2U2N  station wagon                   7       16      24   
2396  8VVXB03.2U2N            SUV                   7       15      22   
2398  8VVXB03.2U2N            SUV                   7       14      20   
2399  8VVXB03.2U2N            SUV                   7       14      20   
2400  8VVXB04.4U2N            SUV                   7       13      19   

     Cmb_MPG Greenhouse_Gas_Score  
0         17                    4  
2         19                    5  
4         19                    5  
6         21                    6  
7         20                    6  
...      ...                  ...  
2394      19                    5  
2396      17                    4  
2398      16                    4  
2399      16                    4  
2400      15                    3  

[1084 rows x 15 columns]

In [13]:
df_18

Model  Displ  Cyl       Trans Drive                  Fuel  \
1       ACURA RDX    3.5  6.0  SemiAuto-6   2WD              Gasoline   
3       ACURA RDX    3.5  6.0  SemiAuto-6   4WD              Gasoline   
4       ACURA TLX    2.4  4.0       AMS-8   2WD              Gasoline   
6       ACURA TLX    3.5  6.0  SemiAuto-9   2WD              Gasoline   
8       ACURA TLX    3.5  6.0  SemiAuto-9   4WD              Gasoline   
...           ...    ...  ...         ...   ...                   ...   
1601  VOLVO XC 60    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   
1603  VOLVO XC 90    2.0  4.0  SemiAuto-8   2WD              Gasoline   
1605  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
1607  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD              Gasoline   
1609  VOLVO XC 90    2.0  4.0  SemiAuto-8   4WD  Gasoline/Electricity   

           Stnd  Underhood_ID     Veh_Class  Air_Pollution_Score City_MPG  \
1            U2  JHNXT03.5GV3     small SUV                    3       20   
3            U2  JHNXT03.5GV3     small SUV                    3       19   
4     L3ULEV125  JHNXV02.4WH3     small car                    3       23   
6     L3ULEV125  JHNXV03.5LH3     small car                    3       20   
8     L3ULEV125  JHNXV03.5LH3     small car                    3       21   
...         ...           ...           ...                  ...      ...   
1601  L3SULEV30  JVVXJ02.0P30     small SUV                    7    26/60   
1603   L3ULEV70  JVVXJ02.0A70  standard SUV                    5       22   
1605   L3ULEV70  JVVXJ02.0A70  standard SUV                    5       22   
1607   L3ULEV70  JVVXJ02.0B70  standard SUV                    5       20   
1609  L3SULEV30  JVVXJ02.0P30  standard SUV                    7    26/63   

     Hwy_MPG Cmb_MPG  Greenhouse_Gas_Score  
1         28      23                     5  
3         27      22                     4  
4         33      27                     6  
6         32      24                     5  
8         30      24                     5  
...      ...     ...                   ...  
1601   28/58   26/59                    10  
1603      29      25                     5  
1605      28      24                     5  
1607      27      23                     5  
1609   30/61   27/62                    10  

[798 rows x 14 columns]

## Drop 'cert_region' column from each df, since our 'cert_region' is implied

In [12]:
df_08 = df_08.drop(columns=['Cert_Region'])
df_18 = df_18.drop(columns=['Cert_Region'])

## Drop records with missing values

In [14]:
df_08 = df_08.dropna()

In [15]:
df_18 = df_18.dropna()

In [16]:
df_18.isnull().sum()

Model                   0
Displ                   0
Cyl                     0
Trans                   0
Drive                   0
Fuel                    0
Stnd                    0
Underhood_ID            0
Veh_Class               0
Air_Pollution_Score     0
City_MPG                0
Hwy_MPG                 0
Cmb_MPG                 0
Greenhouse_Gas_Score    0
dtype: int64

In [17]:
df_08.isnull().sum()

Model                   0
Displ                   0
Cyl                     0
Trans                   0
Drive                   0
Fuel                    0
Stnd                    0
Underhood_ID            0
Veh_Class               0
Air_Pollution_Score     0
City_MPG                0
Hwy_MPG                 0
Cmb_MPG                 0
Greenhouse_Gas_Score    0
dtype: int64

# CHECKPOINT: SAVE EVERYTHING TO .CSV

In [18]:
df_08.to_csv('alpha_08_df.csv')
df_18.to_csv('alpha_18_df.csv')
print('saved')

saved


In [281]:
df_08 = pd.read_csv('alpha_08_df.csv')
df_18 = pd.read_csv('alpha_18_df.csv')

## Fix all datatypes

In [19]:
df_18.dtypes

Model                    object
Displ                   float64
Cyl                     float64
Trans                    object
Drive                    object
Fuel                     object
Stnd                     object
Underhood_ID             object
Veh_Class                object
Air_Pollution_Score       int64
City_MPG                 object
Hwy_MPG                  object
Cmb_MPG                  object
Greenhouse_Gas_Score      int64
dtype: object

In [20]:
df_08.dtypes

Model                    object
Displ                   float64
Cyl                      object
Trans                    object
Drive                    object
Fuel                     object
Stnd                     object
Underhood_ID             object
Veh_Class                object
Air_Pollution_Score      object
City_MPG                 object
Hwy_MPG                  object
Cmb_MPG                  object
Greenhouse_Gas_Score     object
dtype: object

In [21]:
df_08 = df_08.astype({'Cyl':'int64'})
df_18 = df_18.astype({'Air_Pollution_Score':'float64', 'Greenhouse_Gas_Score':'float64', 'Cyl':'int64'})

In [22]:
df_08 = df_08.drop(columns=['Unnamed: 0'])
df_18 = df_18.drop(columns=['Unnamed: 0'])

KeyError: "['Unnamed: 0'] not found in axis"

# CHECKPOINT: SAVE EVERYTHING TO .CSV

In [267]:
df_08.to_csv('alpha_08_df.csv')
df_18.to_csv('alpha_18_df.csv')
print('saved')

saved


## Change datatypes in 'air_pollution' column

In [289]:
df_18

Model  Displ  Cyl       Trans Drive         Fuel       Stnd  \
0             ACURA RDX    3.5    6  SemiAuto-6   2WD     Gasoline         U2   
1             ACURA RDX    3.5    6  SemiAuto-6   4WD     Gasoline         U2   
2             ACURA TLX    2.4    4       AMS-8   2WD     Gasoline  L3ULEV125   
3             ACURA TLX    3.5    6  SemiAuto-9   2WD     Gasoline  L3ULEV125   
4             ACURA TLX    3.5    6  SemiAuto-9   4WD     Gasoline  L3ULEV125   
..                  ...    ...  ...         ...   ...          ...        ...   
830  TOYOTA Sequoia FFV    5.7    8  SemiAuto-6   4WD          Gas   L3ULEV70   
831   TOYOTA Tundra FFV    5.7    8  SemiAuto-6   4WD          Gas   L3ULEV70   
832           VOLVO S90    2.0    4  SemiAuto-8   4WD  Electricity  L3SULEV30   
833         VOLVO XC 60    2.0    4  SemiAuto-8   4WD  Electricity  L3SULEV30   
834         VOLVO XC 90    2.0    4  SemiAuto-8   4WD  Electricity  L3SULEV30   

     Underhood_ID     Veh_Class  Air_Pollution_Score City_MPG Hwy_MPG Cmb_MPG  \
0    JHNXT03.5GV3     small SUV                  3.0       20      28      23   
1    JHNXT03.5GV3     small SUV                  3.0       19      27      22   
2    JHNXV02.4WH3     small car                  3.0       23      33      27   
3    JHNXV03.5LH3     small car                  3.0       20      32      24   
4    JHNXV03.5LH3     small car                  3.0       21      30      24   
..            ...           ...                  ...      ...     ...     ...   
830  JTYXT05.7M58  standard SUV                  5.0       13      17      14   
831  JTYXT05.7M58        pickup                  5.0       13      17      15   
832  JVVXV02.0P30   midsize car                  7.0       70      72      71   
833  JVVXJ02.0P30     small SUV                  7.0       60      58      59   
834  JVVXJ02.0P30  standard SUV                  7.0       63      61      62   

     Greenhouse_Gas_Score  
0                     5.0  
1                     4.0  
2                     6.0  
3                     5.0  
4                     5.0  
..                    ...  
830                   1.0  
831                   2.0  
832                  10.0  
833                  10.0  
834                  10.0  

[835 rows x 14 columns]

## Find records with slashed '/' values and separate them into their own separate records

In [23]:
df_08.Air_Pollution_Score.value_counts()

6      512
7      409
9.5     80
9        7
6/4      1
Name: Air_Pollution_Score, dtype: int64

In [24]:
df_08.Fuel.value_counts()

Gasoline       1007
ethanol/gas       1
CNG               1
Name: Fuel, dtype: int64

In [28]:
df_08.query('City_MPG == "13/18"')

Empty DataFrame
Columns: [Model, Displ, Cyl, Trans, Drive, Fuel, Stnd, Underhood_ID, Veh_Class, Air_Pollution_Score, City_MPG, Hwy_MPG, Cmb_MPG, Greenhouse_Gas_Score]
Index: []

#### Select all rows with records of slashed values

In [25]:
for i in range(len(df_18[df_18['Fuel'].str.contains('/')])):
    flist = df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Fuel'].split('/')
    cmlist = df_18[df_18['Fuel'].str.contains('/')].iloc[i]['City_MPG'].split('/')
    hmlist = df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Hwy_MPG'].split('/')
    blist = df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Cmb_MPG'].split('/')
    #print(f'{flist}{cmlist}{hmlist}{blist}')
    df_18 = df_18.append({
        'Model': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Model'], 
        'Displ': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Displ'], 
        'Cyl': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Cyl'], 
        'Trans': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Trans'], 
        'Drive': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Drive'], 
        'Fuel': flist[1], 
        'Stnd': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Stnd'], 
        'Underhood_ID': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Underhood_ID'], 
        'Veh_Class': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Veh_Class'], 
        'Air_Pollution_Score': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Air_Pollution_Score'],
        'City_MPG': cmlist[1],
        'Hwy_MPG': hmlist[1], 
        'Cmb_MPG': blist[1],
        'Greenhouse_Gas_Score': df_18[df_18['Fuel'].str.contains('/')].iloc[i]['Greenhouse_Gas_Score']
    }, ignore_index=True)
df_18['Fuel'] = df_18['Fuel'].apply(lambda x: x.split('/')[0])
df_18['City_MPG'] = df_18['City_MPG'].apply(lambda x: x.split('/')[0])
df_18['Hwy_MPG'] = df_18['Hwy_MPG'].apply(lambda x: x.split('/')[0])
df_18['Cmb_MPG'] = df_18['Cmb_MPG'].apply(lambda x: x.split('/')[0])

In [344]:
df_08

Model  Displ  Cyl    Trans Drive      Fuel Stnd  \
0              ACURA MDX    3.7    6  Auto-S5   4WD  Gasoline   U2   
1              ACURA RDX    2.3    4  Auto-S5   4WD  Gasoline   U2   
2               ACURA RL    3.5    6  Auto-S5   4WD  Gasoline   U2   
3               ACURA TL    3.2    6  Auto-S5   2WD  Gasoline   U2   
4               ACURA TL    3.5    6  Auto-S5   2WD  Gasoline   U2   
...                  ...    ...  ...      ...   ...       ...  ...   
1005         VOLVO XC 90    3.2    6  Auto-S6   2WD  Gasoline   U2   
1006         VOLVO XC 90    3.2    6  Auto-S6   4WD  Gasoline   U2   
1007         VOLVO XC 90    4.4    8  Auto-S6   4WD  Gasoline   U2   
1008  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD   ethanol   L2   
1009  MERCEDES-BENZ C300    3.0    6  Auto-L7   2WD  Gasoline   L2   

      Underhood_ID    Veh_Class  Air_Pollution_Score  City_MPG  Hwy_MPG  \
0     8HNXT03.7PKR          SUV                  7.0      15.0     20.0   
1     8HNXT02.3DKR          SUV                  7.0      17.0     22.0   
2     8HNXV03.5HKR  midsize car                  7.0      16.0     24.0   
3     8HNXV03.5HKR  midsize car                  7.0      18.0     26.0   
4     8HNXV03.5RKR  midsize car                  7.0      17.0     26.0   
...            ...          ...                  ...       ...      ...   
1005  8VVXB03.2U2N          SUV                  7.0      14.0     20.0   
1006  8VVXB03.2U2N          SUV                  7.0      14.0     20.0   
1007  8VVXB04.4U2N          SUV                  7.0      13.0     19.0   
1008  8MBXV03.0U2A    small car                  6.0      13.0     19.0   
1009  8MBXV03.0U2A    small car                  4.0      18.0     25.0   

      Cmb_MPG  Greenhouse_Gas_Score  
0        17.0                   4.0  
1        19.0                   5.0  
2        19.0                   5.0  
3        21.0                   6.0  
4        20.0                   6.0  
...       ...                   ...  
1005     16.0                   4.0  
1006     16.0                   4.0  
1007     15.0                   3.0  
1008     15.0                   7.0  
1009     21.0                   6.0  

[1010 rows x 14 columns]

In [26]:
cols = ['Fuel', 'City_MPG', 'Hwy_MPG', 'Cmb_MPG', 'Air_Pollution_Score', 'Greenhouse_Gas_Score']
x = df_08[df_08['City_MPG'].str.contains('/')] # copy of dataframe that only contains slashed data

In [27]:
# or a better way
x1 = x.copy()
x2 = x.copy()
for i in cols:
    x1[i] = x1[i].apply(lambda x: x.split('/')[0]) # make x1 only contain the first half of slashed data
    x2[i] = x2[i].apply(lambda x: x.split('/')[1]) # make x2 only contain the second half of the slashed data
df_08 = df_08.drop(x.index) # drop slashed data
df_08 = df_08.append(x1, ignore_index=True) # add unslashed data
df_08 = df_08.append(x2, ignore_index=True) # add unslashed data

### Repeat for 2018

In [29]:
df_18[df_18['Fuel'].str.contains('/')]

Empty DataFrame
Columns: [Model, Displ, Cyl, Trans, Drive, Fuel, Stnd, Underhood_ID, Veh_Class, Air_Pollution_Score, City_MPG, Hwy_MPG, Cmb_MPG, Greenhouse_Gas_Score]
Index: []

## Gas/Gasoline?

In [30]:
df_18['Fuel'] = df_18['Fuel'].apply(lambda x: x.replace('Gas', 'Gasoline'))
df_08['Fuel'] = df_08['Fuel'].apply(lambda x: x.replace('gas', 'Gasoline'))

In [31]:
df_18['Fuel'] = df_18['Fuel'].apply(lambda x: x.replace('Gasolineoline', 'Gasoline'))

## Fix `*_mpg, greenhouse_gas_score and air_pollution_score` datatypes

In [33]:
df_08.dtypes

Model                    object
Displ                   float64
Cyl                       int64
Trans                    object
Drive                    object
Fuel                     object
Stnd                     object
Underhood_ID             object
Veh_Class                object
Air_Pollution_Score     float64
City_MPG                float64
Hwy_MPG                 float64
Cmb_MPG                 float64
Greenhouse_Gas_Score    float64
dtype: object

In [34]:
df_18.dtypes

Model                    object
Displ                   float64
Cyl                       int64
Trans                    object
Drive                    object
Fuel                     object
Stnd                     object
Underhood_ID             object
Veh_Class                object
Air_Pollution_Score     float64
City_MPG                float64
Hwy_MPG                 float64
Cmb_MPG                 float64
Greenhouse_Gas_Score    float64
dtype: object

In [32]:
df_08 = df_08.astype({'Air_Pollution_Score':'float64', 'Greenhouse_Gas_Score':'float64', 'City_MPG':'float64', 'Hwy_MPG':'float64', 'Cmb_MPG':'float'})
df_18 = df_18.astype({'City_MPG':'float64', 'Hwy_MPG':'float64', 'Cmb_MPG':'float'})

# CHECKPOINT: SAVE EVERYTHING

In [35]:
df_08.to_csv('alpha_08_df.csv')
df_18.to_csv('alpha_18_df.csv')
print('saved')

saved


# Create visuals of findings

## Fix `greenhouse_gas_score, air_pollution_score` datatypes

In [38]:
df_08.Fuel.value_counts()
df_18.Fuel.value_counts()

Gasoline       778
Ethanol         26
Diesel          19
Electricity     12
Name: Fuel, dtype: int64

## How many more car models use alternative fuels in 2018 as opposed to 2008?
* ### May need to Google the answer to find out which fuel types are considered 'alternative'

In [41]:
altfuel_08_count = df_08.query('Fuel != "Gasoline"').count()[0]
altfuel_18_count = df_18.query('Fuel != "Gasoline" and Fuel != "Diesel"').count()[0]
additional_altfuel_cars = altfuel_18_count - altfuel_08_count
print(additional_altfuel_cars)

36


## How much has the average fuel economy improved since 2008?

In [42]:
avg_fuel_economy_increase = df_18['Cmb_MPG'].mean() - df_08['Cmb_MPG'].mean()
print(round(avg_fuel_economy_increase, 2))

4.28


## Which type of vehicle's fuel economy has improved the most since 2008?

In [45]:
# MIDSIZE CAR IMPROVED BY:
midcar = df_18.groupby('Veh_Class').Cmb_MPG.mean().round(2)[1] - df_08.groupby('Veh_Class').Cmb_MPG.mean().round(2)[2]
print(midcar.round(2))

6.26


#### (2008) Average fuel economy by vehicle type

In [43]:
df_08.groupby('Veh_Class').Cmb_MPG.mean().round(2)

Veh_Class
SUV              18.47
large car        18.51
midsize car      21.57
minivan          19.12
pickup           16.39
small car        21.06
station wagon    22.37
van              14.83
Name: Cmb_MPG, dtype: float64

#### (2018) Average fuel economy by vehicle type

In [44]:
df_18.groupby('Veh_Class').Cmb_MPG.mean().round(2)

Veh_Class
large car          23.41
midsize car        27.83
minivan            20.80
pickup             18.59
small SUV          24.07
small car          25.42
special purpose    18.50
standard SUV       18.16
station wagon      27.63
Name: Cmb_MPG, dtype: float64

## Which model has the highest `air_pollution_score` and fuel economy in 2008 and 2018? Compare them. 

In [46]:
bestAPS_18 = df_18.iloc[df_18['Air_Pollution_Score'].idxmax()]
bestAPS_08 = df_08.iloc[df_08['Air_Pollution_Score'].idxmax()]
bestFE_18 = df_18.iloc[df_18['Cmb_MPG'].idxmax()]
bestFE_08 = df_08.iloc[df_08['Cmb_MPG'].idxmax()]

In [47]:
print(bestFE_18)

Model                   CHEVROLET Volt
Displ                              1.5
Cyl                                  4
Trans                              CVT
Drive                              2WD
Fuel                       Electricity
Stnd                         L3ULEV125
Underhood_ID              JGMXV01.5020
Veh_Class                    small car
Air_Pollution_Score                  3
City_MPG                           113
Hwy_MPG                             99
Cmb_MPG                            106
Greenhouse_Gas_Score                10
Name: 809, dtype: object


In [48]:
df_18.query('Model == "CHEVROLET Volt"')

Model  Displ  Cyl Trans Drive         Fuel       Stnd  \
226  CHEVROLET Volt    1.5    4   CVT   2WD     Gasoline  L3ULEV125   
227  CHEVROLET Volt    1.5    4   CVT   2WD     Gasoline  L3SULEV30   
809  CHEVROLET Volt    1.5    4   CVT   2WD  Electricity  L3ULEV125   
810  CHEVROLET Volt    1.5    4   CVT   2WD  Electricity  L3SULEV30   

     Underhood_ID  Veh_Class  Air_Pollution_Score  City_MPG  Hwy_MPG  Cmb_MPG  \
226  JGMXV01.5020  small car                  3.0      43.0     42.0     42.0   
227  JGMXV01.5030  small car                  7.0      43.0     42.0     42.0   
809  JGMXV01.5020  small car                  3.0     113.0     99.0    106.0   
810  JGMXV01.5030  small car                  7.0     113.0     99.0    106.0   

     Greenhouse_Gas_Score  
226                  10.0  
227                  10.0  
809                  10.0  
810                  10.0

# This data is strange.
### 2008 best air pollution score is better than 2018 best air pollution score (BMW 328i (9.5) > Audi A3 (7.0))
### 2018 best fuel economy is Chevrolet Volt at 106 MPG.... but the chevy volt is all-electric. So it doesn't use any gas. So.... it shouldn't have an MPG value? or it should be infinite?